pkg/sql/catalog/descs/descriptor.go

```diff
@@ -230,137 +254,233 @@ func (q *byIDLookupContext) lookupLeased(
-			return false, nil, err
+	lookupStoreCacheID := func() (continueOrHalt, descpb.ID, error) {
+		if cd := tc.stored.GetCachedByName(parentID, parentSchemaID, name); cd != nil {
+			return haltLookups, cd.GetID(), nil
 		}
-		desc = ret[0]
-		return desc != nil, desc, err
+		return continueLookups, descpb.InvalidID, nil
 	}
+	lookupLeasedID := func() (continueOrHalt, descpb.ID, error) {
+		if flags.AvoidLeased || flags.RequireMutable || lease.TestingTableLeasesAreDisabled() {
+			return continueLookups, descpb.InvalidID, nil
+		}
+		if isSchema && isTemporarySchema(name) {
+			return continueLookups, descpb.InvalidID, nil
+		}
+		ld, shouldReadFromStore, err := tc.leased.getByName(
+			ctx, tc.deadlineHolder(txn), parentID, parentSchemaID, name,
+		)
+		if err != nil {
+			return haltLookups, descpb.InvalidID, err
+		}
+		if shouldReadFromStore {
+			return continueLookups, descpb.InvalidID, nil
+		}
+		return haltLookups, ld.GetID(), nil
+	}
+	lookupStoredID := func() (continueOrHalt, descpb.ID, error) {
+		id, err := tc.stored.LookupDescriptorID(ctx, txn, parentID, parentSchemaID, name)
+		return haltLookups, id, err
+	}
+
+	// Iterate through each layer until an ID is conclusively found or not, or an
+	// error is thrown.
+	for _, fn := range []func() (continueOrHalt, descpb.ID, error){
+		lookupTemporarySchemaID,
+		lookupSchemaID,
+		lookupSyntheticID,
+		lookupUncommittedID,
+		lookupStoreCacheID,
+		lookupLeasedID,
+		lookupStoredID,
+	} {
+		isDone, id, err := fn()
+		if err != nil {
+			return descpb.InvalidID, err
+		}
+		if isDone {
+			return id, nil
+		}
+	}
+	return descpb.InvalidID, nil
 }
 
 // finalizeDescriptors ensures that all descriptors are (1) properly validated
 // and (2) if mutable descriptors are requested, these are present in the
 // uncommitted descriptors layer.
-// Known validation levels can optionally be provided via validationLevels.
-// If none are provided, finalizeDescriptors seeks them out in the appropriate
-// layer (stored or uncommitted).
-
-// nil safe defaults are used instead. In any case, after validation is
-// performed the known levels are raised accordingly.
 func (tc *Collection) finalizeDescriptors(
 	ctx context.Context,
 	txn *kv.Txn,
 	flags tree.CommonLookupFlags,
 	descs []catalog.Descriptor,
 	validationLevels []catalog.ValidationLevel,
 ) error {
-	if validationLevels == nil {
-		validationLevels = make([]catalog.ValidationLevel, len(descs))
-		for i, desc := range descs {
-			if tc.uncommitted.getUncommittedByID(desc.GetID()) != nil {
-				// Uncommitted descriptors should, by definition, already have been
-				// validated at least at the MutableRead level. This effectively
-				// excludes them from being validated again right now.
-				//
-				// In any case, they will be fully validated when the transaction
-				// commits.
-				validationLevels[i] = validate.MutableRead
-			} else {
-				validationLevels[i] = tc.stored.GetValidationLevelByID(desc.GetID())
-			}
-		}
-	}
```

Comment:
This can go away now that `finalizeDescriptors` is called exclusively from `getDescriptorsByID`. Frankly, this is a relief.